File: src/algorithm/template/detection/base/interpretation.py
from abc import abstractmethod
from typing import Never

from src.algorithm.template.detection.base.detection import (
    TextDetection,
)
from src.algorithm.template.layout.field.base import Field


class BaseInterpretation:
    def __init__(self, text_detection: TextDetection) -> None:
        self.text_detection = text_detection
        self.is_attempted = text_detection is not None
        self.detected_text = text_detection.detected_text if self.is_attempted else ""

    def get_value(self):
        return self.detected_text


class FieldInterpretation:
    def __init__(
        self,
        tuning_config,
        field: Field,
        file_level_detection_aggregates,
        file_level_interpretation_aggregates,
    ) -> None:
        self.tuning_config = tuning_config
        self.field = field
        self.is_attempted = None
        # self.field_block = field.field_block
        self.empty_value = field.empty_value
        self.field_level_confidence_metrics = {}
        # To be updated by child classes
        self.interpretations: list[BaseInterpretation] = []
        # TODO: make get_drawing_instance fetch singleton classes?
        self.drawing = self.get_drawing_instance()

        self.run_interpretation(
            field, file_level_detection_aggregates, file_level_interpretation_aggregates
        )

    @abstractmethod
    def get_drawing_instance(self) -> Never:
        msg = "Not implemented"
        raise Exception(msg)

    @abstractmethod
    def run_interpretation(
        self,
        field,
        file_level_detection_aggregates,
        file_level_interpretation_aggregates,
    ) -> Never:
        msg = "Not implemented"
        raise Exception(msg)

    @abstractmethod
    def get_field_interpretation_string() -> str:
        msg = "Not implemented"
        raise Exception(msg)

    def get_field_level_confidence_metrics(self):
        return self.field_level_confidence_metrics

    def insert_field_level_confidence_metrics(
        self, next_field_level_confidence_metrics
    ) -> None:
        self.field_level_confidence_metrics = {
            **self.field_level_confidence_metrics,
            **next_field_level_confidence_metrics,
        }

File: omrchecker-js/packages/core/src/processors/detection/base/interpretation.ts
/**
 * Interpretation base classes.
 *
 * TypeScript port of src/processors/detection/base/interpretation.py
 * Defines base classes for field interpretation results.
 */

import type { Field } from '../../layout/field/base';
import { TextDetection } from './detection';

/**
 * Base class for single interpretation result.
 *
 * Wraps a TextDetection result with interpretation metadata.
 */
export class BaseInterpretation {
  public textDetection: TextDetection | null;
  public isAttempted: boolean;
  public detectedText: string;

  constructor(textDetection: TextDetection | null) {
    this.textDetection = textDetection;
    this.isAttempted = textDetection !== null;
    this.detectedText = this.isAttempted && textDetection ? textDetection.detectedText || '' : '';
  }

  /**
   * Get the interpreted value.
   *
   * @returns Detected text value
   */
  getValue(): string {
    return this.detectedText;
  }
}

/**
 * Drawing instance interface (forward declaration).
 */
export interface InterpretationDrawing {
  // Stub for now, can be extended when needed
}

/**
 * Abstract base class for field interpretation.
 *
 * Manages interpretation results for a field, including confidence metrics
 * and multiple interpretation results (for multi-marking scenarios).
 */
export abstract class FieldInterpretation {
  public tuningConfig: Record<string, unknown>;
  public field: Field;
  public isAttempted: boolean | null = null;
  public emptyValue: string;
  public fieldLevelConfidenceMetrics: Record<string, unknown> = {};
  public interpretations: BaseInterpretation[] = [];
  public drawing: InterpretationDrawing;

  constructor(
    tuningConfig: Record<string, unknown>,
    field: Field,
    fileLevelDetectionAggregates: unknown,
    fileLevelInterpretationAggregates: unknown
  ) {
    this.tuningConfig = tuningConfig;
    this.field = field;
    this.emptyValue = field.emptyValue;

    // TODO: make get_drawing_instance fetch singleton classes?
    this.drawing = this.getDrawingInstance();

    this.runInterpretation(
      field,
      fileLevelDetectionAggregates,
      fileLevelInterpretationAggregates
    );
  }

  /**
   * Abstract method to get drawing instance.
   * Must be implemented by subclasses.
   */
  abstract getDrawingInstance(): InterpretationDrawing;

  /**
   * Abstract method to run interpretation.
   * Must be implemented by subclasses.
   *
   * @param field - Field to interpret
   * @param fileLevelDetectionAggregates - File-level detection aggregates
   * @param fileLevelInterpretationAggregates - File-level interpretation aggregates
   */
  abstract runInterpretation(
    field: Field,
    fileLevelDetectionAggregates: unknown,
    fileLevelInterpretationAggregates: unknown
  ): void;

  /**
   * Abstract method to get field interpretation string.
   * Must be implemented by subclasses.
   *
   * @returns Final interpretation string
   */
  abstract getFieldInterpretationString(): string;

  /**
   * Get field-level confidence metrics.
   *
   * @returns Confidence metrics
   */
  getFieldLevelConfidenceMetrics(): Record<string, unknown> {
    return this.fieldLevelConfidenceMetrics;
  }

  /**
   * Insert/merge additional field-level confidence metrics.
   *
   * @param nextFieldLevelConfidenceMetrics - Additional metrics to merge
   */
  insertFieldLevelConfidenceMetrics(
    nextFieldLevelConfidenceMetrics: Record<string, unknown>
  ): void {
    this.fieldLevelConfidenceMetrics = {
      ...this.fieldLevelConfidenceMetrics,
      ...nextFieldLevelConfidenceMetrics,
    };
  }
}

